tools/bitfield_gen.md

```diff
@@ -0,0 +1,672 @@
+supports [tagged unions](#tagged-unions) and various layout options for these
+blocks.
+
+## Bit Field Generator by Example
+
+This section walks through the features of the bitfield specification languages
+by example. The [full syntax](#syntax-reference) is described in the next
+section for reference.
+
+### Blocks
+
+The basic concepts of the specification languages are blocks which contain a
+list of fields. The example used in the introduction declares a block with
+name `VMFault`, which contains four fields and two padding entries:
+
+```bf_gen
+block VMFault {
+    field     address           32
+
+    field     FSR               5
+    padding                     7
+    field     instructionFault  1
+    padding                     15
+    field     seL4_FaultType    4
+}
+```
+
+This will generate a type `VMFault_t` in C. Assuming that the word size of the
+underlying architecture is 32 bits, the `VMFault_t` type above will have a size
+of two machine words. The sizes of the content and padding fields in a block
+must add up to a multiple of the machine word size.
+
+The bits in the block are laid out in memory in the order they are mentioned in
+the specification. That is, the left-most 32 bits (i.e. bits 63 to 32) of
+`VMFault` will correspond to the `address` field, the next 5 bits to `FSR`, the
+next 7 bits are inaccessible padding which is set to 0 on creation of the
+bitfield, then one bit for `instructionFault`, 15 bits of padding, and finally 4
+bits (bits 0 to 3 of the entire type) for `seL4_FaultType`.
+
+The following diagram corresponds to the block specification above.
+
+```bf_dia
+   │------------------------  address  ----------------------------│
+   ┌─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┐
+   │                                                               │ ..
+   └─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┘
+   63                                                             32
+
+                     instruction fault                  seL4_FaultType
+   │-- FSR --│             │-|                             │-------|
+   ┌─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┐
+.. │         │             │ │                             │       │
+   └─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┘
+    31      27              19                              3     0
+```
+
+For creating new values of the type `VMFault`, the generator will provide a
+function `VMFault` that takes an argument of the word size of the machine for
+the value of each field -- for the example above, this is `uint32_t`
+
+```c
+VMFault_t VMFault(uint32_t address, uint32_t FSR, uint32_t instructionFault,
+                  uint32_t seL4_FaultType);
+```
+
+Values that do not fit in the bit size of the target field will be cut off
+accordingly, i.e., as if using `val & mask(size)`.
+
+To read the value of each field, the generator will provide the following
+functions:
```

Comment:
This will be covered by a future pruning section, but for now it might be worth mentioning somewhere around here that the generator only provides functions that are used.

Reply:
Yes, reading #842 I think you're right. The problem is that it's a bit more complicated (it's more the build system that does this than the generator), but I'll figure something out.

Reply:
Now done a bit further down in the document.